### ISSUE TITLE
C# generated stubs use nested classes. These are not recommended by .Net guidelines.

### ISSUE THREAD

**Issue by vsubramani** [grpc/grpc#4587]
C# generated stubs use nested classes. These are not recommended by .Net guidelines.

https://msdn.microsoft.com/en-us/library/tdz1bea9(v=vs.71).aspx

**Comment by jtattermusch** [comment]
To be honest I am not sure how much of an issue this is.  The best practice link doesn't really say why it is a bad idea to use nested classes.

Here's my take on it:
All the generated code by gRPC codegen logically belongs to a `service` defined in the proto file. Therefore, we use a outer class that has the same name as the service declaration to indicate which pieces of generated code (i.e. the descriptor for reflection, the generated client stub, the generated server stub and some utility static methods) belong to the service and it also helps to make sure we won't run into naming collisions easily.

Btw,  For C# proto messages we don't generate nested classes (in most cases), but that is because proto messages are first-class members of the .proto file and there is just once generated class for each message.   For each proto service we need to define at least two classes (a client and a server), so generating just one outer class that contains both these classes is consistent with what C# proto library is doing.

Also, a new client is instantiated using `FooService.NewClient(Channel channel)`  (where FooService is the outer class) and usually, one only needs to reference the FooServiceBase (the server side base class) when inheriting from it (which is only a few times in the entire codebase usually), so he burden of referring to the generated classes using a longer name doesn't seem to be playing an important role here. 

To summarize, I think using the nested classes is better for us than not doing so.

CC @jskeet  @chrisdunelm   in case you have some opinions.

**Comment by jskeet** [comment]
It's also worth noting that C# 6 has `using static` which allows nested types to be used without the calling code being long-winded:

```csharp
using static Foo.Bar.ContainingType;
...
NestedType nt = new NestedType();
```

Now the *level* of nesting is unfortunate in protocol buffers, in that we have an extra intermediary layer called `Types` to avoid naming collisions (e.g. between a nested message called `Color` and a property of the containing message called `Color`). There aren't any good version-tolerant alternatives here, that we've been able to come up with.

So basically, I agree with Jan - I think we're doing the right thing here.
